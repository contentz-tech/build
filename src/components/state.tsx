import { createContext, useContext } from "react";
import { jsx } from "@emotion/core";
import { IState } from "../state";

const StateContext = createContext<IState | null>(null);

interface StateProviderProps {
  state: IState;
  children: JSX.Element[] | JSX.Element;
}

function StateProvider({ state, children }: StateProviderProps) {
  return (
    <StateContext.Provider value={state}>{children}</StateContext.Provider>
  );
}

function useState(): IState {
  const state = useContext(StateContext);
  if (state === null)
    throw new Error(
      "Wrap your component in `StateProvider` and pass a valid state."
    );
  return state;
}

export { StateProvider, useState };
