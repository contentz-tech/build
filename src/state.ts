import { IConfig, config as getConfig } from "./getters/config";
import { articles, IArticles } from "./getters/articles";
import { pages, IPages } from "./getters/pages";
import { slides, ISlides } from "./getters/slides";
import { links, ILink } from "./getters/links";
import { statics } from "./getters/statics";
import { i18n } from "./getters/i18n";

interface IState {
  config: IConfig;
  articles: IArticles;
  pages: IPages;
  slides: ISlides;
  links: ILink[];
  statics: string[];
  i18n: any;
}

async function getState(): Promise<IState> {
  const config: IConfig = await getConfig();
  return {
    config,
    articles: await articles(config),
    pages: await pages(config),
    slides: await slides(config),
    links: await links(),
    statics: await statics(),
    i18n: await i18n(config)
  };
}

export { getState, IState };
