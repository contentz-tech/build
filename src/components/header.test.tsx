/** * @jest-environment jsdom */
import "@testing-library/jest-dom/extend-expect";
import { jsx } from "@emotion/core";
import { render } from "@testing-library/react";
import Header from "./header";
import { StateProvider } from "./state";

const baseState = {
  config: { language: "en" },
  articles: {
    order: [],
    byPath: {}
  },
  slides: {
    order: [],
    byPath: {}
  },
  statics: [],
  pages: {
    order: [],
    byPath: {}
  },
  links: [],
  i18n: {
    header: {
      mainLinkDescription: "Go back the home page",
      articles: "Articles",
      links: "Links",
      slides: "Talks",
      resume: "CV"
    }
  }
};

describe("Header", () => {
  test("Empty", () => {
    const { asFragment } = render(
      <StateProvider state={baseState}>
        <Header />
      </StateProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Only Title", () => {
    const { asFragment, getByText } = render(
      <StateProvider
        state={{
          ...baseState,
          config: { ...baseState.config, title: "Contentz" }
        }}
      >
        <Header />
      </StateProvider>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByText(/Contentz/)).toBeInTheDocument();
    expect(getByText(/Contentz/)).toHaveAttribute(
      "title",
      baseState.i18n.header.mainLinkDescription
    );
    expect(getByText(/Contentz/)).toHaveAttribute("href", "/");
  });

  test("Complete", () => {
    const { asFragment, getByText } = render(
      <StateProvider
        state={{
          ...baseState,
          config: {
            ...baseState.config,
            title: "Contentz",
            navigation: [
              { name: "About", path: "/about/" },
              { name: "Books", path: "https://leanpub.com/u/sergiodxa/" }
            ]
          },
          articles: {
            order: ["/articles/test/"],
            byPath: {}
          },
          links: [{ title: "Author Website", url: "https://sergiodxa.com/" }],
          slides: {
            order: ["/slides/test/"],
            byPath: {}
          },
          resume: {
            basics: {}
          }
        }}
      >
        <Header />
      </StateProvider>
    );

    expect(asFragment()).toMatchSnapshot();

    // Title
    expect(getByText(/Contentz/)).toBeInTheDocument();
    expect(getByText(/Contentz/)).toHaveAttribute(
      "title",
      baseState.i18n.header.mainLinkDescription
    );
    expect(getByText(/Contentz/)).toHaveAttribute("href", "/");

    // Content
    expect(getByText(baseState.i18n.header.articles)).toBeInTheDocument();
    expect(getByText(baseState.i18n.header.articles)).toHaveAttribute(
      "target",
      "_self"
    );
    expect(getByText(baseState.i18n.header.articles)).toHaveAttribute(
      "href",
      "/articles/"
    );

    expect(getByText(baseState.i18n.header.links)).toBeInTheDocument();
    expect(getByText(baseState.i18n.header.links)).toHaveAttribute(
      "target",
      "_self"
    );
    expect(getByText(baseState.i18n.header.links)).toHaveAttribute(
      "href",
      "/links/"
    );

    expect(getByText(baseState.i18n.header.slides)).toBeInTheDocument();
    expect(getByText(baseState.i18n.header.slides)).toHaveAttribute(
      "target",
      "_self"
    );
    expect(getByText(baseState.i18n.header.slides)).toHaveAttribute(
      "href",
      "/slides/"
    );

    expect(getByText(baseState.i18n.header.resume)).toBeInTheDocument();
    expect(getByText(baseState.i18n.header.resume)).toHaveAttribute(
      "target",
      "_self"
    );
    expect(getByText(baseState.i18n.header.resume)).toHaveAttribute(
      "href",
      "/cv/"
    );

    // Custom Navigation
    expect(getByText(/About/)).toBeInTheDocument();
    expect(getByText(/About/)).toHaveAttribute("href", "/about/");
    expect(getByText(/About/)).toHaveAttribute("target", "_self");

    expect(getByText(/Books/)).toBeInTheDocument();
    expect(getByText(/Books/)).toHaveAttribute(
      "href",
      "https://leanpub.com/u/sergiodxa/"
    );
    expect(getByText(/Books/)).toHaveAttribute("rel", "nofollow noopener");
    expect(getByText(/Books/)).toHaveAttribute("target", "_blank");
  });
});
