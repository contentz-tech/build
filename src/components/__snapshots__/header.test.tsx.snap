// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Complete 1`] = `
<DocumentFragment>
  .emotion-9 {
  background: white;
  border-bottom: 1px solid #eaeaea;
  box-shadow: 0px 0px 10px 0px rgba(0,0,0,0.12);
  padding: 1em 0;
}

@media print {
  .emotion-9 {
    display: none;
  }
}

.emotion-8 {
  max-width: 60rem;
  margin: 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (max-width:60rem) {
  .emotion-8 {
    padding: 0 1em;
  }
}

.emotion-0 {
  white-space: nowrap;
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-7 {
  overflow-x: auto;
  margin: -1em 0;
  padding: 1em 0;
}

.emotion-7 a {
  padding: 0 1em;
}

.emotion-1 {
  color: #004BCC;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<header
    class="emotion-9"
  >
    <div
      class="emotion-8"
    >
      <a
        class="emotion-0"
        href="/"
        title="Go back the home page"
      >
        Contentz
      </a>
      <nav
        class="emotion-7"
      >
        <a
          class="emotion-1"
          href="/articles/"
          target="_self"
        >
          Articles
        </a>
        <a
          class="emotion-1"
          href="/links/"
          target="_self"
        >
          Links
        </a>
        <a
          class="emotion-1"
          href="/slides/"
          target="_self"
        >
          Talks
        </a>
        <a
          class="emotion-1"
          href="/cv/"
          target="_self"
        >
          CV
        </a>
        <a
          class="emotion-1"
          href="/about/"
          target="_self"
        >
          About
        </a>
        <a
          class="emotion-1"
          href="https://leanpub.com/u/sergiodxa/"
          rel="nofollow noopener"
          target="_blank"
        >
          Books
        </a>
      </nav>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header Empty 1`] = `<DocumentFragment />`;

exports[`Header Only Title 1`] = `
<DocumentFragment>
  .emotion-2 {
  background: white;
  border-bottom: 1px solid #eaeaea;
  box-shadow: 0px 0px 10px 0px rgba(0,0,0,0.12);
  padding: 1em 0;
}

@media print {
  .emotion-2 {
    display: none;
  }
}

.emotion-1 {
  max-width: 60rem;
  margin: 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (max-width:60rem) {
  .emotion-1 {
    padding: 0 1em;
  }
}

.emotion-0 {
  white-space: nowrap;
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<header
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <a
        class="emotion-0"
        href="/"
        title="Go back the home page"
      >
        Contentz
      </a>
    </div>
  </header>
</DocumentFragment>
`;
